### PR TITLE
Feature ResNet, SEResNet

### PR DESCRIPTION
**Pull request by VladislavPatrushev**

### CODE REVIEW

File: src/models/modules/bricks/activations/activations_me.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
+""" Activations (memory-efficient w/ custom autograd)
```
Comment (PososikTeam) [inline src/models/modules/bricks/activations/activations_me.py:1]: Do we really need this functions?

File: src/models/modules/utils/create_attn.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                assert False, 'Invalid attn module (%s)' % attn_type
+        elif isinstance(attn_type, bool):
+            if attn_type:
+                module_cls = SEModule
+        else:
+            module_cls = attn_type
+    return module_cls
+
+
+def create_attn(attn_type, channels, **kwargs):
```
Comment (PososikTeam) [inline src/models/modules/utils/create_attn.py:30]: Documentation

File: src/models/modules/utils/create_attn.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+Adapted from https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/create_attn.py
+Copyright 2019 Ross Wightman
+Licensed under The Apache 2.0 License [see LICENSE for details]
+"""
+import torch
+from src.models.modules.blocks.se import SEModule
+
+
+def get_attn(attn_type):
```
Comment (PososikTeam) [inline src/models/modules/utils/create_attn.py:11]: Documentation

File: src/models/modules/utils/create_attn.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+""""TorchOK Attention Factory.
+
+Adapted from https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/create_attn.py
+Copyright 2019 Ross Wightman
+Licensed under The Apache 2.0 License [see LICENSE for details]
+"""
+import torch
+from src.models.modules.blocks.se import SEModule
```
Comment (PososikTeam) [inline src/models/modules/utils/create_attn.py:8]: Strage things that SE block need in utils. May be replace file?

File: src/models/backbones/utils/hub.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import json
```
Comment (PososikTeam) [inline src/models/backbones/utils/hub.py:1]: I think this module would be rewrited.

File: src/models/backbones/utils/helpers.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
+""" Model creation / weight loading / state_dict helpers
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:1]: I think this file will be overwritten.

File: src/models/backbones/utils/helpers.py [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+    """
+    # Overlay default cfg values from `external_default_cfg` if it exists in kwargs
+    overlay_external_default_cfg(default_cfg, kwargs)
+    # Set model __init__ args that can be determined by default_cfg (if not already passed as kwargs)
+    set_default_kwargs(kwargs, names=('num_classes', 'global_pool', 'in_chans'), default_cfg=default_cfg)
+    # Filter keyword args for task specific model variants (some 'features only' models, etc.)
+    filter_kwargs(kwargs, names=kwargs_filter)
+
+
+def build_model_with_cfg(
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:454]: I think you need rewrite this method...

File: src/models/backbones/utils/registry.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
+import fnmatch
```
Comment (PososikTeam) [inline src/models/backbones/utils/registry.py:1]: Is we really need this?

File: src/models/modules/blocks.py [hunk 9]
```diff
@@ -1 +0,0 @@
-# Higher level modules like SE (Squeeze Excitation)
```
Comment (PososikTeam) [inline src/models/modules/blocks.py:1]: Remove, add comment in __init__.py

File: src/models/modules/bricks.py [hunk 10]
```diff
@@ -1 +1,211 @@
 # Low level modules like ConvBnAct
```
Comment (PososikTeam) [inline src/models/modules/bricks.py:1]: Move to briks directory

File: src/models/modules/bricks.py [hunk 11]
```diff
@@ -1 +1,211 @@
...
+        if (len(weight.shape) != 2 or weight.shape[0] != num_experts or
+                weight.shape[1] != num_params):
+            raise (ValueError(
+                'CondConv variables must have shape [num_experts, num_params]'))
+        for i in range(num_experts):
+            initializer(weight[i].view(expert_shape))
+    return condconv_initializer
+
+
+class CondConv2d(nn.Module):
```
Comment (PososikTeam) [inline src/models/modules/bricks.py:36]: Documentation

File: src/models/modules/bricks.py [hunk 12]
```diff
@@ -1 +1,211 @@
...
+import torch
+from torch import nn as nn
+from torch.nn import functional as F
+
+from .utils.helpers import to_2tuple
+from .conv2d_same import conv2d_same
+from .utils.padding import get_padding_value
+
+
+def get_condconv_initializer(initializer, num_experts, expert_shape):
```
Comment (PososikTeam) [inline src/models/modules/bricks.py:23]: Documentation. And check if it possible to use python @. 
https://realpython.com/primer-on-python-decorators/#:~:text=%40my_decorator%0Adef%20say_whee()%3A%0A%20%20%20%20print(%22Whee!%22)

File: src/models/modules/bricks/conv2d_same.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
+"""TorchOK Conv2d w/ Same Padding
```
Comment (PososikTeam) [inline src/models/modules/bricks/conv2d_same.py:1]: I think we need create new folder for model bricks.

File: src/models/modules/bricks/conv2d_same.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+from typing import Tuple, Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from src.models.modules.utils.padding import pad_same, get_padding_value
+
+
+def conv2d_same(
```
Comment (PososikTeam) [inline src/models/modules/bricks/conv2d_same.py:16]: Documentation.

File: src/models/modules/bricks/conv2d_same.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        x, weight: torch.Tensor, bias: Optional[torch.Tensor] = None, stride: Tuple[int, int] = (1, 1),
+        padding: Tuple[int, int] = (0, 0), dilation: Tuple[int, int] = (1, 1), groups: int = 1):
+    x = pad_same(x, weight.shape[-2:], stride, dilation)
+    return F.conv2d(x, weight, bias, stride, (0, 0), dilation, groups)
+
+
+class Conv2dSame(nn.Conv2d):
```
Comment (PososikTeam) [inline src/models/modules/bricks/conv2d_same.py:23]: Documentation.

File: src/models/modules/bricks/conv2d_same.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1, bias=True):
+        super(Conv2dSame, self).__init__(
+            in_channels, out_channels, kernel_size, stride, 0, dilation, groups, bias)
+
+    def forward(self, x):
+        return conv2d_same(x, self.weight, self.bias, self.stride, self.padding, self.dilation, self.groups)
+
+
+def create_conv2d_pad(in_chs, out_chs, kernel_size, **kwargs):
```
Comment (PososikTeam) [inline src/models/modules/bricks/conv2d_same.py:36]: Documentation.

File: src/models/modules/utils/config.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
+""" Model / Layer Config singleton state
```
Comment (PososikTeam) [inline src/models/modules/utils/config.py:1]: Do we really need this file?

File: src/models/modules/utils/create_act.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
+"""TorchOK Activation Factory.
```
Comment (PososikTeam) [inline src/models/modules/utils/create_act.py:1]: Do we really need this file, if yes replace it in briks path and add documentation.

File: src/models/modules/utils/padding.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
+""" Padding Helpers
```
Comment (PososikTeam) [inline src/models/modules/utils/padding.py:1]: Add documentation.

File: src/models/modules/utils/helpers.py [hunk 20]
```diff
@@ -0,0 +1,31 @@
+"""Layer/Module Helpers
```
Comment (PososikTeam) [inline src/models/modules/utils/helpers.py:1]: Add documentation.

File: src/models/backbones/resnet.py [hunk 21]
```diff
@@ -0,0 +1,343 @@
...
+        """Init Bottleneck.
+
+        Args:
+            in_channels: Number of input channels.
+            out_channels: Number of output channels.
+            stride: Stride.
+            downsample: Downsample module.
+            attn_layer: Attention block.
+        """
+        super(Bottleneck, self).__init__()
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:137]: `super().__init__()`

File: src/models/backbones/resnet.py [hunk 22]
```diff
@@ -0,0 +1,343 @@
...
+        """Init BasicBlock.
+
+        Args:
+            in_channels: Number of input channels.
+            out_channels: Number of output channels.
+            stride: Stride.
+            downsample: Downsample module.
+            attn_layer: Attention block.
+        """
+        super(BasicBlock, self).__init__()
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:83]: `super().__init__()`

File: src/models/backbones/resnet.py [hunk 23]
```diff
@@ -0,0 +1,343 @@
...
+            layers: Number of layers.
+            in_chans: Input channels.
+            block_args: Arguments for block_args.
+        """
+        self.block_args = block_args or dict()
+        self.inplanes = 64
+        self.channels = [64, 128, 256, 512]
+        self.num_features = 512 * block.expansion
+
+        super(ResNet, self).__init__()
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:199]: `super().__init__()`

File: src/models/backbones/resnet.py [hunk 24]
```diff
@@ -0,0 +1,343 @@
...
+
+        self.layer3 = self.__make_layer(block, self.channels[2], layers[2], stride=2, **self.block_args)
+        self._feature_info.append(FeatureInfo(num_channels=self.channels[2], stride=2, module_name='layer3'))
+
+        self.layer4 = self.__make_layer(block, self.channels[3], layers[3], stride=2, **self.block_args)
+        self._feature_info.append(FeatureInfo(num_channels=self.channels[3], stride=2, module_name='layer4'))
+
+        self._create_hooks()
+
+    def __make_layer(self,
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:221]: Documentation

File: src/models/backbones/resnet.py [hunk 25]
```diff
@@ -0,0 +1,343 @@
...
+            block_args: Arguments for block_args.
+        """
+        self.block_args = block_args or dict()
+        self.inplanes = 64
+        self.channels = [64, 128, 256, 512]
+        self.num_features = 512 * block.expansion
+
+        super(ResNet, self).__init__()
+
+        self.conv1 = nn.Conv2d(in_chans, self.channels[0], kernel_size=7, stride=2, padding=3, bias=False)
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:201]: Why not create ConvBnAct in briks?

File: src/models/backbones/utils/helpers.py [hunk 26]
```diff
@@ -0,0 +1,228 @@
+""" TorchOK Model creation / weight loading / state_dict helpers
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:1]: Do we still need all of this file? Can we shorten it?

File: src/models/backbones/utils/helpers.py [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+from typing import Any, Callable, Optional, Tuple, Dict
+
+import torch
+import torch.nn as nn
+from torch.hub import load_state_dict_from_url
+
+_logger = logging.getLogger(__name__)
+
+
+def load_state_dict(checkpoint_path, use_ema=False):
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:22]: Documentation

File: src/models/backbones/utils/helpers.py [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+        else:
+            state_dict = checkpoint
+        _logger.info(f"Loaded {state_dict_key} from checkpoint '{checkpoint_path}'")
+        return state_dict
+    else:
+        _logger.error(f"No checkpoint found at '{checkpoint_path}'")
+        raise FileNotFoundError()
+
+
+def adapt_input_conv(in_chans: int, conv_weight):
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:45]: Documentation

File: src/models/backbones/utils/helpers.py [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+            # NOTE this strategy should be better than random init, but there could be other combinations of
+            # the original RGB input layer weights that'd work better for specific cases.
+            repeat = int(math.ceil(in_chans / 3))
+            conv_weight = conv_weight.repeat(1, repeat, 1, 1)[:, :in_chans, :, :]
+            conv_weight *= (3 / float(in_chans))
+    conv_weight = conv_weight.to(conv_type)
+    return conv_weight
+
+
+def load_pretrained(model: nn.Module,
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:70]: I think this method must be overwritten

File: src/models/backbones/utils/helpers.py [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+
+    if filter_fn is not None:
+        # for backwards compat with filter fn that take one arg, try one first, the two
+        try:
+            state_dict = filter_fn(state_dict)
+        except TypeError:
+            state_dict = filter_fn(state_dict, model)
+
+    input_convs = default_cfg.get('first_conv', None)
+    if input_convs is not None and in_chans != 3:
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:106]: It's good feature to load weights from rgb Model to grayscale Model. It must be save.

File: src/models/backbones/utils/helpers.py [hunk 31]
```diff
@@ -0,0 +1,228 @@
...
+                                 kwargs: Dict):
+    """ Overlay 'external_default_cfg' in kwargs on top of default_cfg arg."""
+    external_default_cfg = kwargs.pop('external_default_cfg', None)
+    if external_default_cfg:
+        default_cfg.pop('url', None)  # url should come from external cfg
+        default_cfg.pop('hf_hub', None)  # hf hub id should come from external cfg
+        default_cfg.update(external_default_cfg)
+
+
+def set_default_kwargs(kwargs: dict,
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:138]: Documentation

File: src/models/backbones/utils/helpers.py [hunk 32]
```diff
@@ -0,0 +1,228 @@
...
+    _logger.info(f'Loading pretrained weights from url ({pretrained_url})')
+    state_dict = load_state_dict_from_url(pretrained_url, progress=progress, map_location='cpu')
+    if 'state_dict' in state_dict:
+        state_dict = state_dict['state_dict']
+
+    if filter_fn is not None:
+        # for backwards compat with filter fn that take one arg, try one first, the two
+        try:
+            state_dict = filter_fn(state_dict)
+        except TypeError:
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:102]: Why need tru except? I think need calls only `state_dict = filter_fn(state_dict)`. And also question is we need filter_fn?

File: src/models/backbones/resnet.py [hunk 33]
```diff
@@ -0,0 +1,342 @@
...
+            out_channels: Number of output channels.
+            stride: Stride.
+            downsample: Downsample module.
+            attn_layer: Attention block.
+        """
+        super().__init__()
+        out_block_channels = out_channels * self.expansion
+
+        self.convbnact = ConvBnAct(in_channels, out_channels, kernel_size=3, padding=1, stride=stride)
+        self.conv = nn.Conv2d(out_channels, out_block_channels, kernel_size=3, padding=1, bias=False)
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:87]: How you look at idea to add functional in ConvBnAct to create briks without activation layer. So you can add ConvBnAct for conv and bn too.

File: src/models/backbones/resnet.py [hunk 34]
```diff
@@ -0,0 +1,342 @@
...
+            downsample: Downsample module.
+            attn_layer: Attention block.
+        """
+        super().__init__()
+        out_block_channels = out_channels * self.expansion
+
+        self.convbnact1 = ConvBnAct(in_channels, out_channels, kernel_size=1, padding=0, stride=1)
+        self.convbnact2 = ConvBnAct(out_channels, out_channels, kernel_size=3, padding=1, stride=stride)
+
+        self.conv = nn.Conv2d(out_channels, out_block_channels, kernel_size=1, stride=1, bias=False)
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:139]: Same thing to create ConvBnAct with possibility to not include activation.

File: src/models/backbones/resnet.py [hunk 35]
```diff
@@ -0,0 +1,342 @@
...
+
+    def __init__(self,
+                 block: Union[BasicBlock, Bottleneck],
+                 layers: List[int],
+                 in_chans: int = 3,
+                 block_args: Dict = None):
+        """Init ResNet.
+
+        Args:
+            block: Type of block.
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:179]: Class?

File: src/models/backbones/resnet.py [hunk 36]
```diff
@@ -0,0 +1,342 @@
...
+            block: Abstraction for convenient work.
+            in_channel: Input channels.
+            blocks: Number of blocks.
+            stride: Stride.
+            block_args: Arguments of block.(for example: attn_layer)
+        """
+        downsample = None
+
+        if stride != 1 or self.inplanes != in_channel * block.expansion:
+            downsample = nn.Sequential(
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:221]: good point about the circle

File: src/models/backbones/resnet.py [hunk 37]
```diff
@@ -0,0 +1,342 @@
...
+                          module_names=['layer1', 'layer2', 'layer3', 'layer4'],
+                          strides=[1, 2, 2, 2])
+
+    def __make_layer(self,
+                     block: Union[BasicBlock, Bottleneck],
+                     in_channel: int,
+                     blocks: int,
+                     stride: int = 1,
+                     **block_args) -> nn.Sequential:
+        """Create layer - abstraction for convenient work.
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:209]: Create Resnet module which contain downsample and one of BasicBlock or Bottleneck modules.

File: src/models/backbones/resnet.py [hunk 38]
```diff
@@ -0,0 +1,342 @@
...
+
+        Args:
+            block: Abstraction for convenient work.
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:212]: Block type class.

File: src/models/backbones/resnet.py [hunk 39]
```diff
@@ -0,0 +1,342 @@
...
+        x = self.maxpool(x)
+
+        x = self.layer1(x)
+        x = self.layer2(x)
+        x = self.layer3(x)
+        x = self.layer4(x)
+
+        return x
+
+    def get_features_info(self, output_channels, module_names, strides):
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:248]: Typing and documentation

File: src/models/backbones/resnet.py [hunk 40]
```diff
@@ -0,0 +1,342 @@
...
+        features_info = []
+        for num_channels, module_name, stride in zip(output_channels, module_names, strides):
+            feature_info = FeatureInfo(module_name=module_name, num_channels=num_channels, stride=stride)
+            features_info.append(feature_info)
+        return features_info
+
+    def get_forward_output_channels(self) -> Union[int, List[int]]:
```
Comment (PososikTeam) [inline src/models/backbones/resnet.py:255]: Typing and documentation

File: src/models/backbones/resnet.py (hunk 36 above)
Comment (PososikTeam) [reply src/models/backbones/resnet.py:214]: Why we still need nn.Sequential?

File: src/models/modules/bricks/convbnact.py [hunk 41]
```diff
@@ -0,0 +1,46 @@
...
+    """Combination of convolution, batchnorm and activation layers."""
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 kernel_size: int,
+                 padding: int,
+                 stride: int = 1,
+                 bias: bool = False,
+                 act_layer: nn.Module = nn.ReLU):
```
Comment (VladVin) [inline src/models/modules/bricks/convbnact.py:16]: Make it optional. And below when it's called just check for None. Using `nn.Identity` isn't a common style

File: src/models/backbones/utils/helpers.py [hunk 42]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    model = model_cls(**model_args)
+    pretrained_url = default_cfg.get('url', None)
+    in_chans = model_args.get('in_chans', 3)
+
+    if pretrained and pretrained_url is not None:
+        state_dict = load_state_dict_from_url(pretrained_url, progress=False, map_location='cpu')
+        input_convs = default_cfg.get('first_conv', None)
+        if input_convs is not None and in_chans != 3:
+            state_dict[input_convs] = adapt_input_conv(in_chans, state_dict[input_convs])
```
Comment (VladVin) [inline src/models/backbones/utils/helpers.py:71]: Remove this

File: src/models/modules/blocks/se.py [hunk 43]
```diff
@@ -0,0 +1,67 @@
...
+        elif reduction_ratio is not None:
+            reduction_channels = self.__make_divisible(channels * reduction_ratio, divisor, min_channels)
+        else:
+            reduction_channels = self.__make_divisible(channels // reduction, divisor, min_channels)
+        self.fc1 = nn.Conv2d(channels, reduction_channels, kernel_size=1, bias=True)
+        self.act = nn.ReLU(inplace=True)
+        self.fc2 = nn.Conv2d(reduction_channels, channels, kernel_size=1, bias=True)
+        self.gate = nn.Sigmoid()
+
+    def __make_divisible(self, value, divisor=8, min_value=None):
```
Comment (VladVin) [inline src/models/modules/blocks/se.py:45]: Put typings

File: tests/models/backbones/test_backbones.py [hunk 44]
```diff
@@ -0,0 +1,68 @@
...
+
+        self.__model = BACKBONES.get('resnet50')(pretrained=True, in_chans=3)
+        self.assertTrue(self.__model(self.__input).equal(self.__output['output_last_layer']))
+
+    def get_output(self, name):
+        def hook(model, input, output):
+            self.__output[name] = output.detach()
+        return hook
+
+# class TestSEResNet50(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (VladVin) [reply tests/models/backbones/test_backbones.py:48]: I would suggest to remove at all, because the code snippet is very similar to the above code

**Review by VladVin — CHANGES_REQUESTED** [review]

**Pull request merged**